### PR TITLE
Aggregation tests for Wildcard Field

### PR DESCRIPTION
**Pull request by not-napoleon**
Add a couple of unit tests for running the Wildcard Field through terms agg and composite terms agg.  I had expected to need to do some additional work to get composite to work, but turns out it just works already, so this is now a test-only PR.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 1]
```diff
@@ -535,13 +537,27 @@ protected void doAssertReducedMultiBucketConsumer(Aggregation agg, MultiBucketCo
...
+
+    protected <T extends AggregationBuilder, V extends InternalAggregation> void testCase(
+        T aggregationBuilder,
+        Optional<IndexWriterConfig> indexWriterConfig, Query query,
+        CheckedConsumer<RandomIndexWriter, IOException> buildIndex,
+        Consumer<V> verify,
+        MappedFieldType... fieldTypes) throws IOException {
         try (Directory directory = newDirectory()) {
-            RandomIndexWriter indexWriter = new RandomIndexWriter(random(), directory);
+            RandomIndexWriter indexWriter;
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:555]: I think Optional prefers things to look like:
```
RandomIndexWriter indexWriter = indexWriterConfg.map(c -> new RandomIndexWriter(random(), directory, c).orElseGet(() -> new RandomIndexWriter(random(), directory);
```

File: x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldAggregationTests.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    private WildcardFieldMapper wildcardFieldMapper;
+    private WildcardFieldMapper.WildcardFieldType wildcardFieldType;
+
+
+    @Before
+    public void setup() {
+        // These settings are lifted out of WildcardFieldMapperTests
+        IndexWriterConfig iwc = newIndexWriterConfig(WildcardFieldMapper.WILDCARD_ANALYZER);
+        iwc.setMergePolicy(newTieredMergePolicy(random()));
+        this.iwc = Optional.of(iwc);
```
Comment (nik9000) [inline x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldAggregationTests.java:47]: This says "when I don't tell you what analyzer to use then use the wildcard analyzer". I believe Elasticsearch always sets the analyzer. The wildcard field mapper looks like it does. I don't think either test needs this.

File: x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldAggregationTests.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            iwc, new MatchAllDocsQuery(),
+            iw -> {
+                indexStrings(iw, "a");
+                indexStrings(iw, "a");
+                indexStrings(iw, "b");
+                indexStrings(iw, "b");
+                indexStrings(iw, "b");
+                indexStrings(iw, "c");
+            },
+            (Consumer<StringTerms>) result -> {
```
Comment (nik9000) [inline x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldAggregationTests.java:97]: `(StringTerms results) ->` should do the same thing with a little less ceremony.

File: x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldAggregationTests.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        );
+
+        testCase(aggregationBuilder,
+            iwc, new MatchAllDocsQuery(),
+            iw -> {
+                indexStrings(iw, "a");
+                indexStrings(iw, "c");
+                indexStrings(iw, "a");
+                indexStrings(iw, "d");
+                indexStrings(iw, "c");
```
Comment (nik9000) [inline x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldAggregationTests.java:126]: Could you try running with `aggregateAfter` set? I think that'd be useful to make sure works as well.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java (hunk 1 above)
Comment (not-napoleon) [reply test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:555]: Yeah, I agree that's what `Optional` would like this to look like.  Unfortunately, `new RandomIndexWriter` can throw an `IOException`, and the lambda `Optional#map` and friends expect is declared to not throw.  I could do something to wrap and throw an unchecked exception, but I think it'd be less clear than just using the if/else block.  Open to suggestions if you know a way to do this that isn't terrible.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 5]
```diff
@@ -547,13 +547,14 @@ protected void doAssertReducedMultiBucketConsumer(Aggregation agg, MultiBucketCo
...
-            MappedFieldType... fieldTypes) throws IOException {
+        T aggregationBuilder,
+        Query query,
+        CheckedConsumer<RandomIndexWriter, IOException> buildIndex,
+        Consumer<V> verify,
+        MappedFieldType... fieldTypes) throws IOException {
         try (Directory directory = newDirectory()) {
-            RandomIndexWriter indexWriter = new RandomIndexWriter(random(), directory);
+            RandomIndexWriter indexWriter;
+            indexWriter = new RandomIndexWriter(random(), directory);
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:557]: Maybe revert this bit too?

**Review by nik9000 — APPROVED** [review]
LGTM. Left a small thing, but otherwise good.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Pull request merged**